### ISSUE TITLE
Background update doesn't check for true internet connectivity before proceeding

### ISSUE THREAD

**Issue by buttercookie42** [breezy-weather/breezy-weather#564]
### Steps to reproduce

1. Leave the phone offline for longer than the background refresh period.
2. Connect to a Wifi network without (immediate) internet access (e.g. due to a captive portal)

### Expected behavior

The update gets deferred until the network actually has true internet connectivity.

### Actual behavior

The update attempts to run anyway, leading to a failed update notification any time I connect to a wifi with a captive portal login.

### Weather source used

AccuWeather

### Breezy Weather version

4.6.11-beta

### Android version

Android 6.0.1

### Device

Moto G4 Play

### Other details

I believe this is because [this bit here](https://github.com/breezy-weather/breezy-weather/blob/main/app/src/main/java/org/breezyweather/background/weather/WeatherUpdateJob.kt#L97C13-L97C13) only checks for [any kind of network connectivity](https://github.com/breezy-weather/breezy-weather/blob/02f171d2ebdadd274ae90e0b583198556639e349/app/src/main/java/org/breezyweather/common/extensions/NetworkExtensions.kt#L27C55-L27C57), but not actual internet access.

I suspect that there should be an additional (or alternative?) check for either `NET_CAPABILITY_INTERNET`, or possible conversely `NET_CAPABILITY_CAPTIVE_PORTAL` *not* being present (not sure whether the latter idea would really successfully work in practice?).

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I read the [Frequently Asked Questions](https://github.com/breezy-weather/breezy-weather/blob/main/HELP.md) document and followed instructions if relevant.
- [X] I have updated the app to **[version 4.6.11 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance), unless my issue only happens with a specific language
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
I understand the issue, but as per Android docs:
- The `NET_CAPABILITY_INTERNET` capability means that the **network is configured** to reach the general Internet.
- `NET_CAPABILITY_CAPTIVE_PORTAL` indicates that this network was found to have a captive portal in place **last time it was probed**.

So, yeah, maybe we could exclude the last one, although I'm not sure how often it probes network.

> I believe this is because [this bit here](https://github.com/breezy-weather/breezy-weather/blob/main/app/src/main/java/org/breezyweather/background/weather/WeatherUpdateJob.kt#L97C13-L97C13) only checks for [any kind of network connectivity](https://github.com/breezy-weather/breezy-weather/blob/02f171d2ebdadd274ae90e0b583198556639e349/app/src/main/java/org/breezyweather/common/extensions/NetworkExtensions.kt#L27C55-L27C57), but not actual internet access.

Actually, it's a second check. The first check is done by Android. When we set up the background job, we ask to only execute it when there is network: https://github.com/breezy-weather/breezy-weather/blob/main/app/src/main/java/org/breezyweather/background/weather/WeatherUpdateJob.kt#L401

If it reaches the 2nd check, it means that Android system considers your network as working and decided to execute the job

**Comment by papjul** [comment]
Interesting discussion, I will read it, there could be some ways of improvements for #441 (although it would not fix the actual loop)

**Comment by papjul** [comment]
It's written that `NET_CAPABILITY_INTERNET` and `NET_CAPABILITY_VALIDATED` do not work with local VPNs such as AdGuard.

Another suggestion includes pinging some websites but that seems overkill and it's not an acceptable option for privacy reasons.

And I don't know how you can check for true connectivity without actually contacting a website to try.

**Comment by buttercookie42** [comment]
Well, in the worst case the VPN problem would remain, but at least for everybody else the situation could be improved by checking for `NET_CAPABILITY_VALIDATED`on top of checking for a physical network connection. Or am I missing something in my logic?

**Comment by papjul** [comment]
Yes, the app would no longer work for people using local VPNs, right?

**Comment by buttercookie42** [comment]
The way I read it, the VPN issue is that in conjunction with a local ad-blocking (or whatever) VPN, `NET_CAPABILITY_VALIDATED` and/or `NET_CAPABILITY_INTERNET` might return true despite no true internet connection being available. In which case things would be no different from today – the update job would proceed and then fail when it actually tries to hit the weather source.

The opposite case would be more worrying, although the discussion only mentions `NET_CAPABILITY_INTERNET` falsely returning false, so maybe `NET_CAPABILITY_VALIDATED` can be relied upon?

**Comment by papjul** [comment]
> Breezy doesnt update in background and same the widget

This issue isn't about background updates not happening, it's the opposite (it happens too often). Please check the troubleshooting section of Settings > Background updates if you have issues with background updates not working.

**Comment by Daniel9132** [comment]
I have the same issue with the notification for a failed refresh appearing every day. 
But I get it every time my phone is in offline mode and it tries to refresh.
I'm using offline mode when I sleep.
Apparently it can't detect offline mode.

Samsung S23+ 
Android 14
Breezy weather 5.3.1

**Comment by papjul** [comment]
> I'm using offline mode when I sleep.

Ensure you turn off ALL networks. This includes VPN and bluetooth.

**Comment by Daniel9132** [comment]
> > Apparently it can't detect offline mode.
> 
> Ensure you turn off ALL networks. This includes VPN and bluetooth.

That's probably it. Im using a VPN based Adblocker with the connection set to always on. It's probably still enabled in offline mode. 
I think it's strange that only breezy weather is affected by this. I have never seen this behavior before and I'm using VPN based Adblockers for many years.

**Comment by papjul** [comment]
Oh no, it's not specific to Breezy Weather, it's an internal function of Android. When I leave VPN on with airplane mode on, I will wake up in the morning with notifications of system update failure to check for updates. If even the system is not able to check for connectivity, our app can't either.

As for the apps that can do it, they probably rely on pinging a website like Google but we don't want it for privacy reasons.

**Comment by Daniel9132** [comment]
> Oh no, it's not specific to Breezy Weather, it's an internal function of Android. When I leave VPN on with airplane mode on, I will wake up in the morning with notifications of system update failure to check for updates. If even the system is not able to check for connectivity, our app can't either.
> 
> As for the apps that can do it, they probably rely on pinging a website like Google but we don't want it for privacy reasons.

Like i mentioned i have never seen this in many years of usage across different devices using VPN Adblockers (mostly Samsung). Maybe it depends on the type of VPN or how it behaves when offline mode is activated.
I'm using a mix of Blokada and Rethink DNS. 

It would be helpful if there was a setting to hide this notification until a certain threshold of hours/days was reached.
Like "hide it for a week if refresh fails".

**Comment by papjul** [comment]
We tell Android to NOT execute our job if there is no network, just like many apps do as it's the standard way to do it. It's up to Android to determine if there is network or not. And Bluetooth and VPN are considered networks by Android.

Other apps may do additional connectivity checks as I said, but for privacy reasons, we don't do it. That's also what GrapheneOS apps do.

Other apps may just silently fail, but we don't do it as the reason for network failure may be different than yours. A network failure because "a VPN is on but without network" is just a small edge case amongst all possible reasons a network can fail. Especially if Android executed our job, meaning there should be network. So we have to report it.

You can either turn off VPN, or choose to hide all "background updates" notifications in the Android app info, but this will include all notifications related to background updates.

**Comment by buttercookie42** [comment]
Checking for `NET_CAPABILITY_VALIDATED` as suggested above before actually attempting to execute the update job (and/or only showing the error message _only_ if the update job fails despite `NET_CAPABILITY_VALIDATED` being present)  still remains as a possible improvement.

**Comment by papjul** [comment]
`NET_CAPABILITY_VALIDATED` returns `true` when using it with a VPN [according to this message](https://gist.github.com/Farbklex/f84029889444ee9c52a331a7e2bd10d2?permalink_comment_id=3888748#gistcomment-3888748), but I see from your initial message that you were having an issue with a captive portal instead

As I don't have any captive portal network to test with, I may attempt to do a testing build for you to test, but this won't solve anything with VPN (which is the most common reported issue)

**Comment by martinhoyer** [comment]
> > Apparently it can't detect offline mode.
> 
> Ensure you turn off ALL networks. This includes VPN and bluetooth.

Oh interesting, thanks for this. I usually switch the phone to airplane mode for the night, but keep bluetooth on (no VPN) and have to dismiss failed update notifications.

**Comment by LegatusVII** [comment]
I get the failed update notifiction when I switch into a user profile with the app after a longer downtime. 
The mobile internet connection is active, but due to using RethinkDNS or a VPN app that activates automaticlly, it hasnt established a real connection yet. There is a lockdown mode without them.
It usually takes a few seconds for it to connect. 

Would the simplest way be just to add a ten second delay once the app notices mobile data/ Wi-Fi is active?
Its not super elegant, but it would probably not show the error then.

**Comment by papjul** [comment]
> Would the simplest way be just to add a ten second delay once the app notices mobile data/ Wi-Fi is active?

Our app is not listening to Internet connection, it's the other way around. The Android system is triggering the background update when and only when there is network available.

Even then, I added a network check as the first step of the background update, to avoid VPN connections with no actual connection. In that case, it will retry work 10 minutes later.

Lockdown mode with no actual connection works fine on my side (= retrying later), so I'm not sure what's the issue in your case.

> I get the failed update notifiction when I switch into a user profile with the app after a longer downtime.

I'm not sure what's your use case, as weather is not really something I would "turn off", but if you put the app in an user profile you only enable on demand, maybe it's best to turn off the background update, and only refresh on demand?

**Comment by LegatusVII** [comment]
Thanks for clarifying for the non-dev. I noticed other apps behave similar, as was stated in the comments earlier too. Difference - they just fail silently. 

I noticed some wireguard configs used in Rethink could lead to servers that are maybe not active or rate limited, I am unsure if I have seen the ten min recheck in action though. Looked at worker.

I occasionally use airplane mode or use a profile just for travel, not necessarily because of tiny security benefit but also to remove distractions. I set the sync to 24hrs, because I just want general weather information, maybe even just for offline. 
But if the sync fails and next one is in 24hrs, at that point I might not be using the profile anymore. 

So I either up the sync frequency or do it manually. Which is not a problem.

I would recommend hiding background services notification by default like other apps handle it. Or you get more people like me asking this? I agree, it is like said niche edge cases, but we might be more vocal. 
I was probably under impression "severe weather and precipitation" alerts run under "background services" notification category too. Or I would have done this earlier.  

Thanks for app!

**Comment by papjul** [comment]
> Difference - they just fail silently.

I noticed that most apps ping Google or NTP services, so if there is no response, they usually skip the network-related features, assuming Google and NTP are never actually offline. But we don't want to do that for privacy reasons.

> I would recommend hiding background services notification by default like other apps handle it.

And then we get tons of reports "I don't understand why it's never refreshing in the background".

And it's not just that, there are many failure reasons. For example, recently, we removed a pollen source that no longer exists. You get refresh errors every time, because you need to take action. Otherwise, your pollen will never display and you might think it's because there is no pollen right now, or the app isn't working correctly. While you actually need to take action.

Error reports are essential for troubleshooting, including failures to connect.

Additionally, in your case, I'm not sure it's a good idea in your case to turn them off. If there is always a network failure, it's best to not run any work at all.

> I was probably under impression "severe weather and precipitation" alerts run under "background services" notification category too. Or I would have done this earlier.

From what I can see, severe weather is in the "Alert" notification channel, and precipitation is not in any group.
But if the background work doesn't run, it won't send anything.
Even then, at a 24 hour refresh rate, alerts and precipitations notifications are not very useful if sent only once a day.

**Comment by lnicola** [comment]
Just ran into this: I think my phone turns off network access overnight. I disabled background service notifications, but it might make sense to dismiss any "update failed" notification after a successful update.

**Comment by papjul** [comment]
> Just ran into this: I think my phone turns off network access overnight.

You can turn on night mode or airplane mode at night, or just turn off Internet. I mean, you don't need to have 5-6 weather updates overnight when you sleep and don't need fresh data. It's better for fair use of weather sources provided to you for free, if you can.

> I disabled background service notifications, but it might make sense to dismiss any "update failed" notification after a successful update.

You will miss all other kind of errors by doing that.

As for dismissing after a successful update, two weather updates may not be equivalent. For example, if you refresh manually in the app and the background update happens to trigger 2 minutes after that, it will be successful, mainly because we will re-used cached data, not because the previous errors were fixed.
